src/drivers/magnetometer/qmc5883p/QMC5883P.cpp

```diff
@@ -0,0 +1,308 @@
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+	}
+    case STATE::CONFIGURE:
+        if (Configure()) {
+            // if configure succeeded then start reading every 20 ms (50 Hz)
+            _state = STATE::READ;
+            ScheduleOnInterval(20_ms, 20_ms);
+
+        } else {
+                // CONFIGURE not complete
+			    if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				    PX4_DEBUG("Configure failed, resetting");
+				    _state = STATE::RESET;
+
+			    } else {
+				    PX4_DEBUG("Configure failed, retrying");
+			    }
+
+            ScheduleDelayed(100_ms);
+        }
+
+        break;
+
+    case STATE::READ: {
+                    struct TransferBuffer{
+                        uint8_t X_LSB;
+				        uint8_t X_MSB;
+				        uint8_t Y_LSB;
+				        uint8_t Y_MSB;
+				        uint8_t Z_LSB;
+				        uint8_t Z_MSB;
+				        uint8_t STATUS;
+                    }buffer{};
+
+                	bool success = false;
+			        uint8_t cmd = static_cast<uint8_t>(Register::X_LSB);
+
+                    if (transfer(&cmd, 1, (uint8_t *)&buffer, sizeof(buffer)) == PX4_OK) {
+                    // process data if successful transfer, no overflow
+				    if ((buffer.STATUS & STATUS_BIT::OVL) == 0) {
+					    int16_t x = combine(buffer.X_MSB, buffer.X_LSB);
+					    int16_t y = combine(buffer.Y_MSB, buffer.Y_LSB);
+					    int16_t z = combine(buffer.Z_MSB, buffer.Z_LSB);
+
+					if (x != _prev_data[0] || y != _prev_data[1] || z != _prev_data[2]) {
+						_prev_data[0] = x;
+						_prev_data[1] = y;
+						_prev_data[2] = z;
+
+						// Sensor orientation(Amovlab_F410)
+						//  Forward X := Z
+						//  Right   Y := Y
+						//  Down    Z := -X
+						int16_t new_x = (z == INT16_MIN) ? INT16_MAX : z;
+						int16_t new_y = (y == INT16_MIN) ? INT16_MAX : y;
+						int16_t new_z = (x == INT16_MIN) ? INT16_MAX : -x;
+
+						_px4_mag.update(now, new_x, new_y, new_z);
+
+						success = true;
+
+						if (_failure_count > 0) {
+							_failure_count--;
+						}
+					}
+				}
+
+			} else {
+				perf_count(_bad_transfer_perf);
+			}
+
+			if (!success) {
+				_failure_count++;
+
+				// full reset if things are failing consistently
+				if (_failure_count > 10) {
+					Reset();
+					return;
+				}
+			}
+
+			if (!success || hrt_elapsed_time(&_last_config_check_timestamp) > 100_ms) {
+				// check configuration registers periodically or immediately following any failure
+				if (RegisterCheck(_register_cfg[_checked_register])) {
+					_last_config_check_timestamp = now;
+					_checked_register = (_checked_register + 1) % size_register_cfg;
+
+				} else {
+					// register check failed, force reset
+					perf_count(_bad_register_perf);
+					Reset();
+				}
+			}
```

Comment:
I am not a fan of doing these periodic reading of register sets in our drivers. It can inadvertently interrupt device operation. The frequent polling of these registers is more likely to cause the exact problem it's trying to prevent. If there is a problem with the device it should get caught quickly in the failure_count logic above.